sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -408,97 +556,217 @@ property_set {
+  // for simplicity we could have just used
+  // kernel_handler LocalKernelHandler = { args_for_kernel_handler_init };
+  // here, but we assume that kernel_handler might be used for more than just
+  // accessing specialization constants and therefore there could be other
+  // initialization parameters which also could be conditional
+  // Even now we don't need to always initialize the kernel_handler object
+  // Re-used body of "sycl_kernel" function:
+  {
+     LocalLambdaClone(LocalKernelHandler);
+   }
+}
+```
+
+As mentioned above, creation of `specialization_constants_buffer` kernel
+argument and initialization of `LocalKernelHandler` object with it only happens
+if we are not targeting SPIR-V, i.e. when we compile code for a target without
+native support for specialization constants.
+
+If that new argument was added, it is communicated to DPC++ through regular
+integration header mechanism, i.e. it is added as new entry to
+`kernel_signatures` structure there with parameter kind set to a new
+enumeration value `kernel_param_kind_t::kind_specialization_constants_buffer`.
+
+Those were descriptions of tasks (1) and (2) of DPC++ FE. Task (3) is to help
+DPC++ RT to connect user-provided `specialization_id` variable with
+corresponding symbolic ID of a specialization constant when
+`handler::set_specialization_constant` is invoked.
+
+As noted above, we can't use regular integration header here, because in general
+case, `specialization_id` variables can't be forward-declared. Therefore, we are
+using integration footer approach, which for the following code snippet:
+```
+struct A {
+  float a, b;
 };
 
-// forward declaration
-extern specialization_id<int> id_int;
+specialization_id<int> id_int;
+specialization_id<A> id_A;
+// ...
+[&](handler &cgh) {
+  cgh.set_specialization_constant<id_int>(42);
+  cgh.set_specialization_constant<id_A>({3.14, 3.14});
+}
+```
+
+Will look like:
+
+```
+namespace detail {
+// generic declaration
+template<auto &SpecName>
+struct get_symbolic_id_helper{};
+
+// specializations for each specialization constant:
+// we can refer to all those specialization_id variables, because integration
+// footer was _appended_ to the user-provided translation unit
+template<>
+struct get_symbolic_id_helper<id_int> {
+  static const char *get_symbolic_id() {
+    return "result of __builtin_unique_ID(id_int) encoded here";
+  }
+};
 
-// specialization
 template<>
-class specialization_constant_info<id_int> {
-  static const char *getName() {
-    return "result of __builtin_unique_stable_name(detail::specialization_id_name_generator<id_int>) encoded here";
+struct get_symbolic_id_helper<id_A> {
+  static const char *get_symbolic_id() {
+    return "result of __builtin_unique_ID(A) encoded here";
   }
 };
+
+} // namespace detail
+
+// TODO: elaborate why we have to include handler implementation here
+#include <CL/sycl/detail/handler.hpp>
```

Comment:
It should be a new header rather than `handler.hpp`

Reply:
Could you please provide a name which can be put here?

Reply:
spec_const_integration.hpp

Reply:
Updated the line in beed9ad0b37e23a28f3cb4c1adb32e70b654e99e